### PR TITLE
create a webhook

### PR DESCRIPTION
**Pull request by TonySchroeder**
close https://github.com/progwise/george-ai/issues/55

WebPageSummary is now a separate collection type in Starpi.

In the "pothos-graphql" codebase, the code has been adapted to the new query, and the type has been defined with a fragment.

The "typesense-cli" continues to fetch data via GraphQL from Strapi. This data is then processed in the "typesense-client" package.

A webhook has been set up in Strapi that is triggered whenever a WebPageSummary is created or updated. The data for this webhook is fetched directly from Strapi, transformed into an object similar to that of GraphQL, and processed in the "typesense-client".

The "typesense-client" package has been created to store the received data in a collection within Typesense.

### CODE REVIEW

File: apps/playwright-scraper/src/main.ts [hunk 1]
```diff
@@ -1,11 +1,11 @@
 import playwright from 'playwright-chromium'
 import { getKeywords, getServiceSummary } from './chat-gpt'
-import { upsertScrapedWebPage } from './strapi.js'
+import { upsertScrapedWebPageAndWebPageSummary } from './strapi.js'
 import { ScrapeResult, scrapePage } from './scrape.js'
 
 const MAX_RUNS = 2 // Maximum number of runs
 
-export interface WebPageSummary extends ScrapeResult {
+export interface ScrapeResultandSummary extends ScrapeResult {
```
Comment (PascalHelbig) [inline apps/playwright-scraper/src/main.ts:8]: `ScrapeResultAndSummary`

File: apps/playwright-scraper/src/main.ts [hunk 2]
```diff
@@ -33,14 +33,14 @@ const processPage = async (url: string): Promise<void> => {
         (await getServiceSummary(scrapeResult.content, language)) ?? ''
       const keywords = (await getKeywords(scrapeResult.content, language)) ?? []
 
-      const webPageSummary: WebPageSummary = {
+      const ScrapeResultandSummary: ScrapeResultandSummary = {
```
Comment (PascalHelbig) [inline apps/playwright-scraper/src/main.ts:36]: `scrapeResultandSummary`

File: apps/playwright-scraper/src/strapi.ts [hunk 3]
```diff
@@ -28,113 +44,182 @@ const CREATE_SCRAPE_WEBPAGE_MUTATION = graphql(`
...
-        OriginalContent: existingScrapedWebPage?.attributes?.OriginalContent,
-        Url: existingScrapedWebPage?.attributes?.Url,
-        WebPageSummaries: summaries,
+    return scrapedWebPage
+  } catch (error) {
+    console.error(error)
+  }
+}
+
+const createOrUpdateWebPageSummary = async (
```
Comment (PascalHelbig) [inline apps/playwright-scraper/src/strapi.ts:160]: rename to upsert

File: apps/playwright-scraper/src/strapi.ts [hunk 4]
```diff
@@ -28,113 +44,182 @@ const CREATE_SCRAPE_WEBPAGE_MUTATION = graphql(`
...
 }
+
+export const upsertScrapedWebPageAndWebPageSummary = async (
+  scrapeResultAndSummary: ScrapeResultandSummary,
+) => {
+  const createdScrapedWebPage = await getOrCreateScrapedWebPage(
+    scrapeResultAndSummary,
+  )
+
+  if (createdScrapedWebPage && createdScrapedWebPage.id) {
```
Comment (PascalHelbig) [inline apps/playwright-scraper/src/strapi.ts:219]:
```suggestion
  if (createdScrapedWebPag?.id) {
```

File: apps/strapi-cms/src/api/web-page-summary/content-types/web-page-summary/lifecycles.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import { upsertTypesenseDocument } from "@george-ai/typesense-client";
```
Comment (PascalHelbig) [inline apps/strapi-cms/src/api/web-page-summary/content-types/web-page-summary/lifecycles.ts:1]: prettier?

File: apps/typesense-cli/src/rebuild.collection.ts [hunk 6]
```diff
@@ -13,98 +12,42 @@ const strapiClient = new GraphQLClient(endpoint, {
...
-      ) || []
-
-    for (const document of documents) {
-      await client.collections(collectionName).documents().upsert(document)
-      console.log(
-        `Data added to typesense in collection ${collectionName}`,
-        document,
-      )
-    }
+    rebuildTypesenseCollection(webPageSummaries)
```
Comment (PascalHelbig) [inline apps/typesense-cli/src/rebuild.collection.ts:47]: For loop?

File: packages/typesense-client/.env.example [hunk 7]
```diff
@@ -0,0 +1,5 @@
+STRAPI_API_KEY=
```
Comment (PascalHelbig) [inline packages/typesense-client/.env.example:1]: file can be deleted?

File: packages/typesense-client/.eslintignore [hunk 8]
```diff
@@ -0,0 +1 @@
+gql
```
Comment (PascalHelbig) [inline packages/typesense-client/.eslintignore:1]: remove file

File: packages/typesense-client/package.json [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+    "typecheck": "tsc --noEmit",
+    "lint": "eslint ."
+  },
+  "dependencies": {
+    "typesense": "^1.7.1"
+  },
+  "devDependencies": {
+    "eslint": "^8.46.0",
+    "eslint-config-custom": "*",
+    "prettier": "^3.0.1",
```
Comment (PascalHelbig) [inline packages/typesense-client/package.json:19]:
```suggestion
```

File: packages/typesense-client/src/index.ts [hunk 10]
```diff
@@ -0,0 +1,2 @@
+export * from './upsert.collection.js'
+export * from './../src/rebuild.collection.js'
```
Comment (PascalHelbig) [inline packages/typesense-client/src/index.ts:2]: why?

File: packages/typesense-client/src/rebuild.collection.ts [hunk 11]
```diff
@@ -0,0 +1,64 @@
+import { ensureCollectionExists, upsertDocument } from './typesense.js'
+
```
Comment (PascalHelbig) [inline packages/typesense-client/src/rebuild.collection.ts:2]: maybe Codegen?

File: packages/typesense-client/.eslintignore (hunk 8 above)
Comment (TonySchroeder) [reply packages/typesense-client/.eslintignore:1]: is needed now

**Comment by moncapitaine** [comment]
Bug: Scraped using wrong locale, locale for the summary must be the one specified in chatGPT prompt.

File: apps/playwright-scraper/src/strapi.ts [hunk 12]
```diff
@@ -16,6 +12,26 @@ const client = new GraphQLClient(endpoint, {
   },
 })
 
+// TODO: should the locale be stored in the ScrapedWebPage or WebPageSummaries?
+const GET_SCRAPE_WEBPAGES_BY_URL_QUERY = graphql(`
```
Comment (moncapitaine) [inline apps/playwright-scraper/src/strapi.ts:16]: single use constant, omit const and add it directly where it is used

File: apps/strapi-cms/src/api/web-page-summary/content-types/web-page-summary/schema.json [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+    },
+    "Keywords": {
+      "type": "string",
+      "pluginOptions": {
+        "i18n": {
+          "localized": true
+        }
+      }
+    },
+    "Summary": {
```
Comment (moncapitaine) [inline apps/strapi-cms/src/api/web-page-summary/content-types/web-page-summary/schema.json:31]: Lowercase for attributes, please correct in all Strapi schema

File: apps/playwright-scraper/src/strapi.ts [hunk 14]
```diff
@@ -16,125 +12,205 @@ const client = new GraphQLClient(endpoint, {
...
-    const existingScrapedWebPage = scrapedWebPages?.data?.[0]
 
-    if (!existingScrapedWebPage?.id) {
-      await client.request(CREATE_SCRAPE_WEBPAGE_MUTATION, {
-        data: {
-          Title: webPageSummary.title,
-          OriginalContent: webPageSummary.content,
-          Url: webPageSummary.url,
-          WebPageSummaries: [newSummary],
+    const webPageSummariesId = webPageSummaries?.data.at(0)?.id
```
Comment (PascalHelbig) [inline apps/playwright-scraper/src/strapi.ts:133]:
```suggestion
    const webPageSummaryId = webPageSummaries?.data.at(0)?.id
```

File: apps/playwright-scraper/src/strapi.ts [hunk 15]
```diff
@@ -16,125 +12,205 @@ const client = new GraphQLClient(endpoint, {
...
-      return
-    }
-    const summaries: Maybe<ComponentWebPageSummaryWebPageSummaryInput>[] =
-      existingScrapedWebPage?.attributes?.WebPageSummaries || []
-    const oldSummaryIndex = summaries.findIndex(
-      (summary) =>
-        summary?.LargeLanguageModel === webPageSummary.largeLanguageModel,
-    )
+      )
+      const createdSummaryId = updateWebPageSummary?.data?.id
```
Comment (PascalHelbig) [inline apps/playwright-scraper/src/strapi.ts:161]:
```suggestion
```

File: packages/typesense-client/src/index.ts [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+        webPageSummary.attributes?.scraped_web_pages?.data?.attributes
+          ?.originalContent,
+      publicationState: webPageSummary.attributes?.scraped_web_pages?.data
+        ?.attributes?.publishedAt
+        ? 'published'
+        : 'draft',
+      keywords: webPageSummary?.attributes?.keywords
+        ? JSON.parse(webPageSummary.attributes?.keywords)
+        : [],
+      summary: webPageSummary.attributes?.summary,
```
Comment (PascalHelbig) [inline packages/typesense-client/src/index.ts:63]: Create an interface and use this interface for the argument of `upsertTypsenseCollection`

File: codegen.ts [hunk 17]
```diff
@@ -5,6 +5,7 @@ const projects = [
   'pothosGraphql',
   'nextjsWeb',
   'typesenseCli',
+  'typesenseClient',
```
Comment (PascalHelbig) [inline codegen.ts:8]: Can be removed

File: graphql.config.ts [hunk 18]
```diff
@@ -65,6 +65,25 @@ module.exports = {
...
+        codegen: {
+          hooks: { afterOneFileWrite: ['prettier --write'] },
+          generates: {
+            'packages/typesense-client/src/gql/': {
+              preset: 'client',
+            },
+          },
+        },
+      },
+    },
```
Comment (PascalHelbig) [inline graphql.config.ts:86]: Please remove again

File: apps/strapi-cms/src/api/web-page-summary/content-types/web-page-summary/lifecycles.ts [hunk 19]
```diff
@@ -0,0 +1,48 @@
+import { upsertTypesenseCollection } from '@george-ai/typesense-client'
+
+const transformWebPageSummary = (webPageSummaryResult) => {
```
Comment (PascalHelbig) [inline apps/strapi-cms/src/api/web-page-summary/content-types/web-page-summary/lifecycles.ts:3]: move into `transformAndUpsertSummary`

File: packages/pothos-graphql/src/WebPageSummary/index.ts [hunk 20]
```diff
@@ -0,0 +1,107 @@
+import { builder } from '../builder'
+import { FragmentType, graphql, useFragment } from '../gql'
+import { GraphQLClient } from 'graphql-request'
+import dotenv from 'dotenv'
+import { WebPageSummaryFragment } from '../gql/graphql'
+
+dotenv.config()
```
Comment (PascalHelbig) [inline packages/pothos-graphql/src/WebPageSummary/index.ts:7]: do we need this?

File: packages/pothos-graphql/src/WebPageSummary/index.ts [hunk 21]
```diff
@@ -0,0 +1,107 @@
...
+
+const endpoint = 'http://localhost:1337/graphql'
+const client = new GraphQLClient(endpoint, {
```
Comment (PascalHelbig) [inline packages/pothos-graphql/src/WebPageSummary/index.ts:10]: Create a separate file `strapiGraphqlClient.ts` and export this as `strapiGraphqlClient`

File: packages/typesense-client/src/index.ts [hunk 22]
```diff
@@ -0,0 +1,77 @@
...
+
+const baseCollectionSchema: {
+  name: string
+  fields: {
+    name: string
+    type: FieldType
+    optional?: boolean
+  }[]
+} = {
+  name: '',
```
Comment (PascalHelbig) [inline packages/typesense-client/src/index.ts:13]: name is missing

File: packages/typesense-client/src/index.ts [hunk 23]
```diff
@@ -0,0 +1,77 @@
...
+    { name: 'keywords', type: 'string[]' },
+    { name: 'summary', type: 'string' },
+    { name: 'largeLanguageModel', type: 'string' },
+  ],
+}
+
+export const upsertTypesenseCollection = async (
+  webPageSummary: WebPageSummaryEntity,
+) => {
+  const collectionName = 'scraped_web_pages_summaries'
```
Comment (PascalHelbig) [inline packages/typesense-client/src/index.ts:30]:
```suggestion
```

use instead `baseCollectionSchema.name`

File: packages/typesense-client/src/index.ts [hunk 24]
```diff
@@ -0,0 +1,77 @@
+import { WebPageSummaryEntity } from './gql/graphql.js'
+import { typesenseClient } from './typesense.js'
+import { FieldType } from 'typesense/lib/Typesense/Collection.js'
+
+const baseCollectionSchema: {
```
Comment (PascalHelbig) [inline packages/typesense-client/src/index.ts:5]: maybe `summaryCollectionSchema` as name instead?

**Review by PascalHelbig — APPROVED** [review]

**Pull request merged**